### ISSUE TITLE
GAE\VGAE to restore graph features

### ISSUE THREAD

**Issue by roman-vygon** [rusty1s/pytorch_geometric#1062]
## ❓ Questions & Help

Good time of the day to you!
I'm trying to use GAE or VGAE architecture to encode and decode features of the graph, not its adjacency, like so (GAE example):
```python
class Encoder(torch.nn.Module):
    def __init__(self, in_channels=None, out_channels=None):
        super(Encoder, self).__init__()
        self.out_channels = out_channels
        self.conv1 = GCNConv(in_channels,2 * out_channels)
        self.conv2 = GCNConv(2 * out_channels, out_channels)
    
    def forward(self, data):        
        x = data.x
        edge_index = data.edge_index
        x = F.relu(self.conv1(x, edge_index))
        x = self.conv2(x, edge_index)    
        return x
    
class Decoder(torch.nn.Module):        
    def __init__(self):
        super(Decoder, self).__init__()        
        self.conv = GCNConv(1, 8)
        self.conv1 = GCNConv(8, 2)
        
    def forward(self, x, edge_index):        
        x = F.relu(self.conv(x, edge_index))
        x = self.conv1(x, edge_index)
        return x

train_loader = DataLoader(dataset_list, batch_size=1, shuffle=True)

encoder = Encoder(in_channels=2, out_channels=1)
model = GAE(encoder, Decoder()).to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
criterion = nn.MSELoss()

def train(batch):    
    model.train()
    optimizer.zero_grad()    
    z = model.encode(batch)            
    pred = model.decoder(z, batch.edge_index)         
    loss = criterion(pred, batch.x)
    loss.backward()
    optimizer.step()
    return loss

iteration = 0
for epoch in range(10):    
    model.train()
    for batch in train_loader: 
        mse_loss = train(batch)
        iteration+=1
```
I tried a similar VGAE architecture, with minor changes to this code, of course.
Howerer, my dataset has a lot of zeros, so the autoencoder tends to output the same number for each feature, something close to zero, adding more epochs or changing the learning rate doesnt help.

1. What architecture would you suggest for autoencoding features of the graph, am I correct to just use some GCNConv layers?
2. How can I deal with this imbalance of zeros in my dataset?

**Comment by rusty1s** [comment]
Hi, overall `GCNConv` may not be well-suited for this task since it cannot really maintain central node information and instead behaves more like a low-pass filter. It is therefore hard for the model to reconstruct high frequencies.

I suggest to look into the `GraphConv` for this task (with `mean` aggregation) that leverages neighborhood aggregation with a simple skip connection.

**Comment by roman-vygon** [comment]
> Hi, overall `GCNConv` may not be well-suited for this task since it cannot really maintain central node information and instead behaves more like a low-pass filter. It is therefore hard for the model to reconstruct high frequencies.
> 
> I suggest to look into the `GraphConv` for this task (with `mean` aggregation) that leverages neighborhood aggregation with a simple skip connection.

Thank you, this improved the accuracy greatly
